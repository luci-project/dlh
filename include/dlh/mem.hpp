// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#pragma once

#include <dlh/types.hpp>
#include <dlh/math.hpp>

namespace Memory {
/*! \brief Allocate a memory block
 * The memory is not initialized.
 *
 * \param size Requested size of memory in bytes.
 * \return Address of memory or `0` on error (no memory available) or if size
 *         was zero.
 */
uintptr_t alloc(size_t size);

template<typename T>
inline T* alloc(size_t size) {
	return reinterpret_cast<T*>(alloc(size));
}

/*! \brief Free an allocated memory block
 *
 * \param addr Address of an previously allocated memory block.
 */
void free(uintptr_t addr);

template<typename T>
inline void free(T* addr) {
	return free(reinterpret_cast<uintptr_t>(addr));
}


/*! \brief Change the size of an allocated memory block
 * The contents will be unchanged in the range from the start of the region up
 * to the minimum of the old and new sizes. If the new size is larger than the
 * old size, the added memory will not be initialized.
 *
 * \param addr Address of an previously allocated memory block.
 *             If `0`, then the call is equivalent to \ref alloc().
 * \param size New size of the memory block. If equal to zero, then the call
 *             is equivalent to free
 * \return Address of new memory block or `0` on error
 */
uintptr_t realloc(uintptr_t addr, size_t size);

template<typename T>
inline T* realloc(T* addr, size_t size) {
	return reinterpret_cast<T*>(realloc(reinterpret_cast<uintptr_t>(addr), size));
}

/*! \brief Allocate memory for an array of elements
 * The memory is set to zero.
 *
 * \param nmemb Number of elements
 * \param size Size of an element in bytes
 * \return address of the allocated memory or `0` if the request fails
 */
uintptr_t alloc_array(size_t nmemb, size_t size);

template<typename T>
inline T* alloc_array(size_t nmemb) {
	return reinterpret_cast<T*>(alloc_array(nmemb, sizeof(T)));
}

/*! \brief Copy a memory area
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return Address of destination
 * \note The memory must not overlap!
 */
inline uintptr_t copy(uintptr_t dest, uintptr_t src, size_t size) {
	unsigned char * __restrict__ destination = reinterpret_cast<unsigned char*>(dest);
	unsigned char const * __restrict__ source = reinterpret_cast<unsigned char const*>(src);

	for(size_t i = 0; i != size; ++i) {
		destination[i] = source[i];
	}

	return dest;
}

/*! \brief Copy a memory area
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return pointer to destination
 * \note The memory must not overlap!
 */
template<typename T, typename U>
inline T* copy(T * __restrict__ dest, U const * __restrict__ src, size_t size) {
	return reinterpret_cast<T*>(copy(reinterpret_cast<uintptr_t>(dest), reinterpret_cast<uintptr_t>(src), size));
}

/*! \brief Copy a memory area
 * while the source may overlap with the destination
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return Address of destination
 */
inline uintptr_t move(uintptr_t dest, uintptr_t src, size_t size) {
	unsigned char * __restrict__ destination = reinterpret_cast<unsigned char*>(dest);
	unsigned char const * __restrict__ source = reinterpret_cast<unsigned char const*>(src);

	if(source > destination) {
		for(size_t i = 0; i != size; ++i)
			destination[i] = source[i];
	} else {
		for(size_t i = size; i != 0; --i)
			destination[i-1] = source[i-1];
	}

	return dest;
}

/*! \brief Copy a memory area
 * while the source may overlap with the destination
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return pointer to destination
 */
template<typename T, typename U>
inline T* move(T * dest, U const * src, size_t size) {
	return reinterpret_cast<T*>(move(reinterpret_cast<uintptr_t>(dest), reinterpret_cast<uintptr_t>(src), size));
}

/*! \brief Fill a memory area with a pattern
 * \ingroup string
 * \param dest destination buffer
 * \param pattern single byte pattern
 * \param size number of bytes to fill with pattern
 * \return address of destination
 */
inline uintptr_t set(uintptr_t dest, int pattern, size_t size) {
	unsigned char *destination = reinterpret_cast<unsigned char*>(dest);

	for(size_t i = 0; i != size; ++i) {
		destination[i] = static_cast<unsigned char>(pattern);
	}

	return dest;
}

/*! \brief Fill a memory area with a pattern
 * \ingroup string
 * \param dest destination buffer
 * \param pattern single byte pattern
 * \param size number of bytes to fill with pattern
 * \return pointer to destination
 */
template<typename T>
inline T* set(T* dest, int pattern, size_t size) {
	return reinterpret_cast<T*>(set(reinterpret_cast<uintptr_t>(dest), pattern, size));
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 pointer to first element
 * \param s2 pointer to second element
 * \param size number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
inline int compare(uintptr_t s1, uintptr_t s2, size_t n) {
	const unsigned char * c1 = reinterpret_cast<const unsigned char*>(s1);
	const unsigned char * c2 = reinterpret_cast<const unsigned char*>(s2);

	for(size_t i = 0; i != n; ++i)
		if (c1[i] != c2[i])
			return static_cast<int>(c1[i]) - static_cast<int>(c2[i]);

	return 0;
}


/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 pointer to first element
 * \param s2 pointer to second element
 * \param size number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T * s1, const U * s2, size_t n) {
	return compare(reinterpret_cast<uintptr_t>(s1), reinterpret_cast<uintptr_t>(s2), n);
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 pointer to first element
 * \param s2 pointer to second element
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T * s1, const U * s2) {
	int r = compare(s1, s2, Math::min(sizeof(T), sizeof(U)));
	return r == 0 ? sizeof(U) - sizeof(T) : r;
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 reference to first element
 * \param s2 reference to second element
 * \param size number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T & s1, const T & s2, size_t n) {
	return compare(reinterpret_cast<uintptr_t>(&s1), reinterpret_cast<uintptr_t>(&s2), n);
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 reference to first element
 * \param s2 reference to second element
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T & s1, const U & s2) {
	int r = compare(reinterpret_cast<uintptr_t>(&s1), reinterpret_cast<uintptr_t>(&s2), Math::min(sizeof(T), sizeof(U)));
	return r == 0 ? sizeof(U) - sizeof(T) : r;
}

}  // namespace Memory
