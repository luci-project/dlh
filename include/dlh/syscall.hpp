#pragma once

#include <dlh/error.hpp>
#include <dlh/types.hpp>
#include <dlh/systypes.hpp>

// Handy helpers for errors
void warn(const char * msg);
void warn(const char * msg, const char * detail);
[[noreturn]] void die(const char * msg = nullptr, const char * detail = nullptr);

/*! \brief Syscalls with errno in return structure */
namespace Syscall {

/*! \brief Return Value with Errno */
template<typename T>
class ReturnValue {
	T _v;
	Errno _e;

 public:
	ReturnValue() : _e(ENONE) {}
	ReturnValue(Errno e) :  _v((T)(-1)), _e(e) {}
	ReturnValue(T v, Errno e = ENONE) :  _v(v), _e(e) {}
	ReturnValue(const ReturnValue& other) : _v(other._v), _e(other._e) {}
	ReturnValue(ReturnValue&& other) = default;

	inline T* operator->() { return &_v; }
	inline const T* operator->() const { return &_v; }
	inline T& operator*() & { return _v; }
	inline const T& operator*() const & { return _v; }
	inline bool success() const { return _e == ENONE; }
	inline bool failed() const { return _e != ENONE; }
	inline Errno error() const { return _e; }
	inline const char * error_message() const { return __errno_string(_e); }
	inline void warn_on_error(const char * msg = nullptr) const { if (!success()) warn(msg, error_message()); }
	inline void exit_on_error(const char * msg = nullptr) const { if (!success()) die(msg, error_message()); }
	inline void die_on_error(const char * msg = nullptr) const { if (!success()) die(msg, error_message()); }
	inline T& value() & { return _v; }
	inline const T& value() const & { return _v; }
	inline T& value_or_exit(const char * msg = nullptr) & { exit_on_error(msg); return value(); }
	inline const T& value_or_exit(const char * msg = nullptr) const & { exit_on_error(msg); return value(); }
	inline T& value_or_die(const char * msg = nullptr) & { die_on_error(msg); return value(); }
	inline const T& value_or_die(const char * msg = nullptr) const & { die_on_error(msg); return value(); }
	inline T& value_or_default(const T& v) & { return success() ? value() : v; }
	inline const T& value_or_default(const T& v) const & { return success() ? value() : v; }
	inline operator bool() const { return success(); }
	template<typename O> ReturnValue & operator=(const ReturnValue<O>& other) { _e = other._e; _v = other._v; return *this; }
	ReturnValue & operator=(const T& other) { _v = other; return *this; }
	template<typename O> bool operator==(const ReturnValue<O>& other) const { return _e == other._e && _v == other._v; }
	template<typename O> bool operator==(const O& other) const { return _e == ENONE && _v == other; }
	template<typename O> bool operator!=(const ReturnValue<O>& other) const { return _e != other._e && _v != other._v; }
	template<typename O> bool operator!=(const O& other) const { return _e != ENONE || _v != other; }
};


ReturnValue<int> nanosleep(const struct timespec *req, struct timespec *rem);
ReturnValue<unsigned> sleep(unsigned seconds);

pid_t gettid();
pid_t getpid();
pid_t getppid();

ReturnValue<int> getrlimit(rlimit_t resource, struct rlimit *rlim);
ReturnValue<int> arch_prctl(arch_code_t code, unsigned long addr);
ReturnValue<int> prctl(prctl_t option, unsigned long arg2, unsigned long arg3 = 0, unsigned long arg4 = 0, unsigned long arg5 = 0);

ReturnValue<int> sigaction(int sig, const struct sigaction * __restrict__ sa, struct sigaction * __restrict__ old);
ReturnValue<int> raise(signal_t sig);

[[noreturn]] void abort();
[[noreturn]] void crash();
[[noreturn]] void exit(int code);

ReturnValue<uintptr_t> mmap(uintptr_t start, size_t len, int prot, int flags, int fd, long off);
ReturnValue<uintptr_t> mremap(uintptr_t old_addr, size_t old_len, size_t new_len, int flags, uintptr_t new_addr = 0);
ReturnValue<int> mprotect(uintptr_t start, size_t len, int prot);
ReturnValue<int> munmap(uintptr_t start, size_t len);
ReturnValue<int> msync(uintptr_t start, size_t len, int flags);

ReturnValue<int> access(const char *filename, int amode);
ReturnValue<ssize_t> readlink(const char * __restrict__ path, char * __restrict__ buf, size_t bufsize);
ReturnValue<int> open(const char *filename, int flags, int mode = 0);
ReturnValue<ssize_t> read(int fd, void *buf, size_t count);
ReturnValue<ssize_t> write(int fd, const void *buf, size_t size);
ReturnValue<int> close(int fd);
ReturnValue<int> fcntl(int fd, fcntl_cmd_t cmd, unsigned long arg = 0);
ReturnValue<int> fallocate(int fd, int mode, off_t base, off_t len);
ReturnValue<int> ftruncate(int fd, off_t length);

ReturnValue<int> fstat(int fd, struct stat *st);
ReturnValue<int> stat(const char * __restrict__ path, struct stat * __restrict__ buf);
ReturnValue<int> lstat(const char * __restrict__ path, struct stat * __restrict__ buf);

ReturnValue<int> userfaultfd(int flags);
ReturnValue<int> ioctl(int fd, int req, void * arg = nullptr);
ReturnValue<int> memfd_create(const char *name, unsigned flags);

ReturnValue<int> inotify_init(int flags = 0);
ReturnValue<int> inotify_add_watch(int fd, const char *pathname, uint32_t mask);
ReturnValue<int> inotify_rm_watch(int fd, int wd);

ReturnValue<int> futex(int * __restrict__ uaddr, futex_op_t futex_op, int val, const void * __restrict__ timeout, int * __restrict__ uaddr2, int val3);

ReturnValue<int> brk(uintptr_t addr);
ReturnValue<uintptr_t> sbrk(intptr_t inc);
}  // namespace Syscall
