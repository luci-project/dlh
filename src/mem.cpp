// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#include <dlh/mem.hpp>
#include <dlh/log.hpp>
#include <dlh/math.hpp>
#include <dlh/mutex.hpp>
#include <dlh/assert.hpp>
#include <dlh/string.hpp>

#include "alloc_buddy.hpp"

namespace Memory {

#define PAGE_SIZE 0x1000UL

#ifndef MMAP_ALLOC_THREASHOLD
// Threshold to use mmap instead of alloc
#define MMAP_ALLOC_THREASHOLD (32 * PAGE_SIZE)
#endif

#ifndef MMAP_ALLOC_START
// Start address for the mmap base allocator
#define MMAP_ALLOC_START (0x700000000000UL)
#endif

#ifndef MIN_ALLOC_LOG2
// Minimum size to  alloc
// Default: 2^10 = 1024 Bytes
#define MIN_ALLOC_LOG2 10
#endif

#ifndef MAX_ALLOC_LOG2
// Maximum total space to alloc
// Default: 2^30 = 1GB
#define MAX_ALLOC_LOG2 30
#endif

#ifdef DLH_LEGACY
// Use program break based allocator (with 128 KiB chunks)
static uintptr_t reserve_sbrk(size_t size, uintptr_t & max_ptr) {
	const size_t block_size = 128 * 1024;
	size = Math::align_up(size, block_size);


	if (auto sbrk = Syscall::sbrk(size)) {
		uintptr_t ptr = sbrk.value();
		assert(max_ptr == 0 || max_ptr == ptr);
		max_ptr = ptr + size;
		return ptr;
	} else {
		LOG_WARNING << "Allocator reserve by incrementing program break of " << size << " bytes failed: " << sbrk.error_message() << endl;
		return 0;
	}
}
static Allocator::Buddy<MIN_ALLOC_LOG2, MAX_ALLOC_LOG2, reserve_sbrk> allocator;
#else
// Use mmap based allocator
static uintptr_t reserve_mmap(size_t size, uintptr_t & max_ptr) {
	// Mapping address
	static uintptr_t mmap_addr = MMAP_ALLOC_START;

	uintptr_t ptr = 0;

	// Calculate size uising 1 MiB chunks
	const size_t block_size = 256 * PAGE_SIZE;
	size = Math::align_up(size, block_size);

	if (max_ptr == 0) {
		// Initial mapping
		if (auto mmap = Syscall::mmap(mmap_addr, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0)) {
			mmap_addr = ptr = mmap.value();
			max_ptr = ptr + size;
		} else {
			LOG_WARNING << "Allocator reserve by mmap of " << size << " bytes at " << reinterpret_cast<void*>(mmap_addr) << " failed: " << mmap.error_message() << endl;
		}
		return ptr;
	} else {
		// Extend mapping

		/* In case we don't have mremap:
		if (auto mmap = Syscall::mmap(max_ptr, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED_NOREPLACE, 0, 0)) {
			ptr = mmap.value();
			assert(max_ptr == ptr);
			max_ptr += size;
		} else {
			LOG_WARNING << "Allocator reserve by mmap of " << size << " bytes at " << (void*)max_ptr << " failed: " << mmap.error_message() << endl;
		}
		*/

		size_t old_size = max_ptr - mmap_addr;
		size_t new_size = old_size + size;
		assert(old_size % block_size == 0);
		if (auto mremap = Syscall::mremap(mmap_addr, old_size, new_size, 0)) {
			assert(mmap_addr == mremap.value());
			ptr = max_ptr;
			max_ptr += size;
		} else {
			LOG_WARNING << "Allocator reserve by mremap of " << old_size << " bytes to " << new_size << " bytes at " << reinterpret_cast<void*>(mmap_addr) << " failed: " << mremap.error_message() << endl;
		}
	}

	return ptr;
}

static Allocator::Buddy<MIN_ALLOC_LOG2, MAX_ALLOC_LOG2, reserve_mmap> allocator;
#endif

static Mutex mutex;

#ifndef DLH_LEGACY
#define MEMORYMAP_MAGIC 0xDEADBEAFBADF00DUL
class MemoryMap {
	size_t size;
	unsigned long magic;
	char data[];

 public:
	static uintptr_t create(size_t size) {
		size += sizeof(MemoryMap);
		if (auto mmap = Syscall::mmap(0, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0)) {
			MemoryMap * info = reinterpret_cast<MemoryMap *>(mmap.value());
			info->magic = MEMORYMAP_MAGIC;
			info->size = size;
			auto addr = reinterpret_cast<uintptr_t>(info->data);
			// Ensure GLIBC compatibility
			assert(addr % 16 == 0);
			return addr;
		} else {
			LOG_WARNING << "Allocator using mmap of " << size << " bytes failed: " << mmap.error_message() << endl;
			return 0;
		}
	}

	static uintptr_t resize(uintptr_t addr, size_t new_size) {
		new_size += sizeof(MemoryMap);
		uintptr_t start = addr - __builtin_offsetof(MemoryMap, data);
		MemoryMap * old_info = reinterpret_cast<MemoryMap *>(start);
		assert(old_info->magic == MEMORYMAP_MAGIC);
		if (auto mremap = Syscall::mremap(start, old_info->size, new_size, MREMAP_MAYMOVE)) {
			MemoryMap * new_info = reinterpret_cast<MemoryMap *>(mremap.value());
			assert(new_info->magic == MEMORYMAP_MAGIC);
			new_info->size = new_size;
			return reinterpret_cast<uintptr_t>(new_info->data);
		} else {
			LOG_WARNING << "Allocator using mremap of " << old_info->size << " bytes to " << new_size << " bytes at " << reinterpret_cast<void*>(start) << " failed: " << mremap.error_message() << endl;
			return 0;
		}
	}

	static void remove(uintptr_t addr) {
		uintptr_t start = addr - __builtin_offsetof(MemoryMap, data);
		MemoryMap * info = reinterpret_cast<MemoryMap *>(start);
		assert(info->magic == MEMORYMAP_MAGIC);
		size_t size = info->size;
		auto munmap = Syscall::munmap(start, size);
		if (munmap.failed()) {
			LOG_WARNING << "Allocator using munmap of " << size << " bytes at " << reinterpret_cast<void*>(start) << " failed: " << munmap.error_message() << endl;
		}
	}
};
#endif

uintptr_t alloc(size_t size) {
	if (size == 0) {
		return 0;
#ifndef DLH_LEGACY
	} else if (size >= MMAP_ALLOC_THREASHOLD) {
		return MemoryMap::create(size);
#endif
	} else {
		Guarded section(mutex);

		uintptr_t addr = allocator.malloc(size);

		// Ensure GLIBC compatibility
		assert(addr % 16 == 0);

		return addr;
	}
}

void free(uintptr_t addr) {
	if (addr != 0) {
#ifndef DLH_LEGACY
		if (allocator.size(addr) == 0) {
			MemoryMap::remove(addr);
		} else  // NOLINT
#endif
		 {
			Guarded section(mutex);

			allocator.free(addr);
		 }
	}
}

uintptr_t realloc(uintptr_t addr, size_t size) {
	if (addr == 0 && size == 0)
		return 0;

	size_t old_size = allocator.size(addr);
	uintptr_t new_addr = 0;

#ifndef DLH_LEGACY
	if ((addr != 0 && old_size == 0) || size >= MMAP_ALLOC_THREASHOLD) {
		if (size == 0) {
			MemoryMap::remove(addr);
		} else if (old_size == 0 && addr != 0) {
			new_addr = MemoryMap::resize(addr, size);
		} else {
			Guarded section(mutex);
			// First try resize in allocator (cheap)
			if (allocator.resize(addr, size)) {
				return addr;
			} else if ((new_addr = MemoryMap::create(size)) != 0 && addr != 0) {
				// Copy contents from allocator to mapping
				copy(new_addr, addr, Math::min(size, old_size));
				// Remove old memory
				allocator.free(addr);
			}
		}
	} else  // NOLINT
#endif
	 {
		assert(addr == 0 || old_size != 0);
		Guarded section(mutex);

		// Allocate (of size > 0)
		if (size > 0) {
			if (old_size != 0 && allocator.resize(addr, size))
				return addr;
			else if ((new_addr = allocator.malloc(size)) != 0 && addr != 0)
				// Copy contents
				copy(new_addr, addr, Math::min(size, old_size));
		}

		// Free old allocation
		if (addr != 0)
			allocator.free(addr);
	 }

	// Ensure GLIBC compatibility
	assert(new_addr % 16 == 0);
	return new_addr;
}

uintptr_t alloc_array(size_t nmemb, size_t size) {
	// Catch integer-multiplication overflow
	size_t bytes = nmemb * size;
	if (size == 0 || bytes / size != nmemb)
		return 0;

#ifndef DLH_LEGACY
	if (bytes >= MMAP_ALLOC_THREASHOLD)
		return MemoryMap::create(bytes);
#endif

	// Initialize memory
	uintptr_t new_addr = alloc(bytes);
	if (new_addr != 0)
		set(new_addr, 0, bytes);

	return new_addr;
}
}  // namespace Memory
