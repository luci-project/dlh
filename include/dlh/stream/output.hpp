#pragma once

#include <dlh/syscall.hpp>
#include <dlh/stream/string.hpp>

template<size_t BUFFERSZ>
class OutputStream : public StringStream<BUFFERSZ> {
 protected:
	int fd;

 public:
	/*! \brief Default constructor  */
	explicit OutputStream(int fd) : fd(fd) {}

	/*! \brief Destructor */
	virtual ~OutputStream() {}

	/*! \brief Flush the buffer.
	 */
	virtual void flush() override {
		size_t l = 0;
		while (l < this->_pos) {
			auto r = Syscall::write(fd, this->_bufptr + l, this->_pos - l);
			if (!r.valid())
				break;
			else
				l += r.value();
		}
		this->_pos = 0;
	}

};

extern OutputStream<1024> cout, cerr;
