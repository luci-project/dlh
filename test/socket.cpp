#include <dlh/stream/output.hpp>
#include <dlh/syscall.hpp>
#include <dlh/mem.hpp>

struct sockaddr_un server_addr = { AF_UNIX, "test_socket.server" };
struct sockaddr_un client_addr = { AF_UNIX, "test_socket.client" };

static void server() {
	if (auto fd = Syscall::socket(PF_UNIX, SOCK_DGRAM, 0)) {
		if (auto bind = Syscall::bind(fd.value(), (struct sockaddr *)&server_addr, sizeof(server_addr))) {
			while (true) {
				char buf[513];
				struct sockaddr_un from;
				socklen_t fromlen = sizeof(from);
				if (auto len = Syscall::recv(fd.value(), buf, 512, 0, (struct sockaddr *)&from, &fromlen)) {
					if (len.value() == 0) {
						cout << "Server recv EOF" << endl;
						break;
					} else {
						buf[len.value()] = 0;
						cout << "Server recv: " << buf << endl;
						for (ssize_t i = 0; i < len.value() ; i++) {
							if ((buf[i] >= 'a' && buf[i] <= 'm') || (buf[i] >= 'A' && buf[i] <= 'M')) {
								buf[i] += 13;
							} else if ((buf[i] >= 'n' && buf[i] <= 'z') || (buf[i] >= 'N' && buf[i] <= 'Z')) {
								buf[i] -= 13;
							}
						}
						if (auto slen = Syscall::send(fd.value(), buf, len.value(), 0, (struct sockaddr *)&from, fromlen)) {
							if (slen.value() != len.value()) {
								cerr << "Server send incomplete: " << slen.value() << " vs " << len.value() << endl;
							}
						} else {
							cerr << "Server send failed: " << slen.error_message() << endl;
						}
					}
				} else {
					cerr << "Server recv failed: " << len.error_message() << endl;
				}
			}
		} else {
			cerr << "Server bind socket failed: " << bind.error_message() << endl;
		}
		Syscall::close(fd);
	} else {
		cerr << "Server creating socket failed: " << fd.error_message() << endl;
	}
}

static void client() {
	if (auto fd = Syscall::socket(PF_UNIX, SOCK_DGRAM, 0)) {
		if (auto bind = Syscall::bind(fd.value(), (struct sockaddr *)&client_addr, sizeof(client_addr))) {
			if (auto connect = Syscall::connect(fd.value(), (struct sockaddr *)&server_addr, sizeof(server_addr))) {
				char buf[] = "Hello world!";
				if (auto len = Syscall::send(fd.value(), buf, sizeof(buf), 0)) {
					if (len.value() != static_cast<int>(sizeof(buf))) {
						cerr << "Client send incomplete: " << len.value() << " vs " << sizeof(buf) << endl;
					} else {
						if (auto rlen = Syscall::recv(fd.value(), buf, 512, 0)) {
							buf[sizeof(buf) - 1] = '\0';
							cout << "Client recv: " << buf << endl;
						} else {
							cerr << "Client recv failed: " << rlen.error_message() << endl;
						}
					}
				} else {
					cerr << "Client send failed: " << len.error_message() << endl;
				}
			} else {
				cerr << "Client connect to socket failed: " << connect.error_message() << endl;
			}
		} else {
			cerr << "Client bind socket failed: " << bind.error_message() << endl;
		}
		Syscall::close(fd);
	} else {
		cerr << "Client creating socket failed: " << fd.error_message() << endl;
	}
}

int main() {
	/*Mem::set(&server_addr, 0, sizeof(server_addr));
	server_addr.sun_family = AF_UNIX;
	strcpy(server_addr.sun_path, server_socket);*/
	Syscall::unlink(server_addr.sun_path);
	Syscall::unlink(client_addr.sun_path);

	if (auto pid = Syscall::fork()) {
		if (pid == 0) {
			server();
		} else {
			Syscall::sleep(1);
			client();
		}
	} else {
		cerr << "Fork failed: " << pid.error_message() << endl;
	}

	return 0;
}
