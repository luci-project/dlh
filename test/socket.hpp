#include <dlh/stream/output.hpp>
#include <dlh/syscall.hpp>

static void rot13(char * buf, ssize_t len) {
	for (ssize_t i = 0; i < len ; i++) {
		if ((buf[i] >= 'a' && buf[i] <= 'm') || (buf[i] >= 'A' && buf[i] <= 'M')) {
			buf[i] += 13;
		} else if ((buf[i] >= 'n' && buf[i] <= 'z') || (buf[i] >= 'N' && buf[i] <= 'Z')) {
			buf[i] -= 13;
		}
	}
}

template<typename SOCKADDR>
static void datagram_server(int socket, SOCKADDR server_addr) {
	if (auto bind = Syscall::bind(socket, reinterpret_cast<struct sockaddr *>(&server_addr), sizeof(server_addr))) {
		while (true) {
			char buf[513];
			SOCKADDR from;
			socklen_t fromlen = sizeof(from);
			if (auto len = Syscall::recv(socket, buf, 512, 0, reinterpret_cast<struct sockaddr *>(&from), &fromlen)) {
				if (len.value() == 0) {
					cout << "Server recv EOF" << endl;
					break;
				} else {
					buf[len.value()] = 0;
					cout << "Server recv: " << buf << endl;
					rot13(buf, len.value());
					if (auto slen = Syscall::send(socket, buf, len.value(), 0, reinterpret_cast<struct sockaddr *>(&from), fromlen)) {
						if (slen.value() != len.value()) {
							cerr << "Server send incomplete: " << slen.value() << " vs " << len.value() << endl;
						}
					} else {
						cerr << "Server send failed: " << slen.error_message() << endl;
					}
				}
			} else {
				cerr << "Server recv failed: " << len.error_message() << endl;
			}
		}
	} else {
		cerr << "Server bind socket failed: " << bind.error_message() << endl;
	}
}

template<typename SOCKADDR>
static void stream_server(int socket, SOCKADDR server_addr) {
	if (auto bind = Syscall::bind(socket, reinterpret_cast<struct sockaddr *>(&server_addr), sizeof(server_addr))) {
		if (auto listen = Syscall::listen(socket, 23)) {
			SOCKADDR client_addr;
			socklen_t client_addrlen = sizeof(client_addr);
			if (auto accept = Syscall::accept(socket, reinterpret_cast<struct sockaddr *>(&client_addr), &client_addrlen)) {
				while (true) {
					char buf[513];
					if (auto len = Syscall::recv(accept.value(), buf, 512, 0)) {
						if (len.value() == 0) {
							cout << "Server recv EOF" << endl;
							break;
						} else {
							buf[len.value()] = 0;
							cout << "Server recv: " << buf << endl;
							rot13(buf, len.value());
							if (auto slen = Syscall::send(accept.value(), buf, len.value(), 0)) {
								if (slen.value() != len.value()) {
									cerr << "Server send incomplete: " << slen.value() << " vs " << len.value() << endl;
								}
							} else {
								cerr << "Server send failed: " << slen.error_message() << endl;
							}
						}
					} else {
						cerr << "Server recv failed: " << len.error_message() << endl;
						break;
					}
				}
			} else {
				cerr << "Server accept failed: " << accept.error_message() << endl;
			}
		} else {
			cerr << "Server listen failed: " << listen.error_message() << endl;
		}
	} else {
		cerr << "Server bind socket failed: " << bind.error_message() << endl;
	}
}

template<typename SOCKADDR>
static void client(int socket, SOCKADDR server_addr) {
	if (auto connect = Syscall::connect(socket, reinterpret_cast<struct sockaddr *>(&server_addr), sizeof(server_addr))) {
		char buf[] = "Hello world!";
		if (auto len = Syscall::send(socket, buf, sizeof(buf), 0)) {
			if (len.value() != static_cast<int>(sizeof(buf))) {
				cerr << "Client send incomplete: " << len.value() << " vs " << sizeof(buf) << endl;
			} else {
				if (auto rlen = Syscall::recv(socket, buf, 512, 0)) {
					buf[sizeof(buf) - 1] = '\0';
					cout << "Client recv: " << buf << endl;
				} else {
					cerr << "Client recv failed: " << rlen.error_message() << endl;
				}
			}
		} else {
			cerr << "Client send failed: " << len.error_message() << endl;
		}
	} else {
		cerr << "Client connect to socket failed: " << connect.error_message() << endl;
	}
}
