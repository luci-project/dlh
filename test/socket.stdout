Server recv: Hello world!
Client recv: Uryyb jbeyq!
